For my final project I plan to utilize the modular toolkit Starish ( https://github.com/egluckthaler/starfish ) to analyze a fungal genome. I already have a high quality assembled genome of a fungal species of interest, and I plan to use Starfish to analyze the genome for the presence of starship elements, which are large mobile genetic elements that can carry accessory genes and contribute to genome evolution.  

The Starfish toolkit is comprised of several different components that can be used to identify, annotate, and analyze starship elements in genomic data. I plan to follow the user guide directly and follow the exact recommendations of the github repository. 

I will start by pulling the apptainer image into my OSC environment, and then follow the steps provided from there using Conda. I have previously attempted to use this toolkit, before having much understanding of how to use a command line interface, so I am hopeful that with my new skills I will be able to successfully run the toolkit and analyze my fungal genome. 

One thing that I am unclear about at this moment from the rubric is "Has individual scripts that are not overly long and donâ€™t do multiple unrelated things". I am not sure if this means that I need to write my own scripts to use the toolkit, or if I can just use the scripts provided in the Starfish toolkit. I will write my own scripts where necessary to run the toolkit, but I am hoping that I can primarily use the scripts provided in the Starfish repository. Even with the suite of tools and user manual outlining the scripts and steps to take, I have previously found its execution to be very challenging, so I'm hoping that it will be acceptable to use the toolkit as is, and that the installation of all the components and dependencies and running the scripts will show sufficient effort and understanding of the material. For the project I will be documeting all the steps that I take and any troubleshooting that I have to do along the way.

Due to the complex nature of this toolkit, I am not sure yet how far I will get in its execution with potential troubleshooting and various steps, but my hope is to complete the entire process and have the genome analyzed for starship elements.

As for the reason why I chose this as my final project, as I've mentioned above, I have already attempted (and failed) to use this toolkit in the past, so I am motivated to try again now that I have more experience with command line interfaces and using bioinformatics tools. Additionally, I am very interested in fungal genomics and mobile genetic elements, so this project aligns well with my research interests. I hope to be able to report on any interesting findings when I might one day be able to describe this new fungal species. Finally, I think that successfully using this toolkit will be a valuable skill to have for my future research endeavors, as mobile genetic elements seem to play a significant role in genome evolution and understanding them can provide insights into various biological processes.     

I am really looking forward to working on this project and learning more about being able to follow complex bioinformatics toolkits and github repository instructions. I believe that this project will not only enhance my technical skills but also deepen my understanding of fungal genomics and mobile genetic elements.